src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,418 @@
+module Algebra.Graph.Acyclic.AdjacencyMap (
+  -- * Data types and type aliases
+  PartialOrder,
+
+  -- * Data structure
+  AdjacencyMap,
+
+  -- * Internal Checks
+  consistent,
+
+  -- * Graph properties
+  isEmpty, edgeList, vertexList, edgeSet, vertexSet,
+  vertexCount, edgeCount, adjacencyList, hasEdge, hasVertex,
+
+  -- * Graph transformation
+  removeVertex, removeEdge, transpose, induce,
+
+  -- * Basic graph construction primitives
+  empty, vertex, overlayD, connectD, vertices,
+
+  -- * Additional functions on acyclic graphs
+  box, topSort, transitiveClosure,
+
+  -- * Acyclic graph construction methods
+  scc, fromGraph
+  ) where
+
+import Algebra.Graph (Graph, foldg)
+import qualified Algebra.Graph.AdjacencyMap as AM
+import qualified Algebra.Graph.AdjacencyMap.Algorithm as AM
+import qualified Algebra.Graph.NonEmpty.AdjacencyMap as NonEmpty
+import qualified Data.Graph.Typed as Typed
+import Data.Set (Set)
+import Data.Coerce (coerce)
+
+{-| The 'AdjacencyMap' data type represents an acyclic graph by
+wrapping around 'Algebra.Graph.AdjacencyMap.AdjacencyMap'. The
+idea is that the user of the library should not be given a way
+to produce a cyclic graph.
+
+We define a 'Num' instance as a convenient notation for working
+with graphs:
+
+  > edgeList   0           == []
+  > vertexList 0           == [0]
+  > edgeList   (1 + 2)     == []
+  > vertexList (1 + 2)     == [1,2]
+  > edgeList   (1 * 2)     == [(1,2)]
+  > vertexList (1 * 2)     == [1,2]
+  > edgeList   (1 + 2 * 3) == [(2,3)]
+  > vertexList (1 + 2 * 3) == [1,2,3]
+  > edgeList   (1 * 2 + 3) == [(1,2)]
+  > vertexList (1 * 2 + 3) == [1,2,3]
+
+The default partial order for the 'Num' instance to restrict
+edges is (\<), ie. Edges can only be formed from vertex
+numbered /a/ to a vertex numbered /b/ if and only if 
+/a/ \< /b/:
+
+  > edgeList (1 * 2) == [(1,2)]
+  > edgeList (2 * 1) == []
+  > (2 * 1) == (2 + 1)
+
+__Note:__ the 'Num' instance does not satisfy several "customary laws"
+of 'Num', which dictate that 'fromInteger' @0@ and 'fromInteger' @1@
+should act as additive and multiplicative identities, and 'negate' as
+additive inverse. Nevertheless, overloading 'fromInteger', '+' and '*'
+is very convenient when working with algebraic graphs; we hope that
+in future Haskell's Prelude will provide a more fine-grained class
+hierarchy for algebraic structures, which we would be able to utilise
+without violating any laws.
+-}
+newtype AdjacencyMap a = AAM
+  { aam :: AM.AdjacencyMap a
+  } deriving (Show, Eq, Ord)
+
+-- | Check if the internal graph representation is consistent,
+-- i.e. that all edges refer to existing vertices and the graph
+-- is acyclic. It should be impossible to create an inconsistent 
+-- adjacency map.
+--
+-- @
+-- consistent 'empty'           == True
+-- consistent (1 + 2)         == True
+-- consistent (1 * 2 + 2 * 3) == True
```

Comment:
These examples do not agree with the testsuite. Just list here everything that you test. 

In general, all examples should be tested.

Reply:
Ah! I've used this initially but changed the test suite to use arbitrary graph instances.

Reply:
Just to avoid a misunderstanding: better change these examples to match the current tests. It's very important to test invariants using `Arbitrary` instance.